### PR TITLE
Fixing multisize and stored requests in AMP Issue #345

### PR DESCRIPTION
**Pull request by pm-deepak-dhage**
PubMatic adslot parameter was accepting string in the format "adslotName@widthxheight". The behavior is changed as follows -
1. adslot is not a mandatory parameter anymore. PubMatic selects default tag to monetise the request without adslot.
2. If present, it can have format "adslotName@widthxheight" or just "adslotName". Size is no more mandatory in the adslot.
3. Format array which was ignored earlier is now passed as it is in the request.
4. If size is present in adslot string, it will be considered else first size from the format (ad_units.sizes) array will be picked as primary size. Either of adslot name or format array must have size in the request.

The PR not only fixes the issue#345, it also makes the adslot handling in sync with Prebid Server written in Go-Lang.

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java [hunk 1]
```diff
@@ -64,27 +64,41 @@ protected Imp modifyImp(Imp imp, ExtImpPubmatic extImpPubmatic) throws PreBidExc
         if (imp.getAudio() != null) {
             modifiedImp.audio(null);
         }
-        if (imp.getBanner() != null) {
-            final String adSlotString = extImpPubmatic.getAdSlot().trim();
-            final String[] adSlot = adSlotString.split("@");
-            if (adSlot.length != 2 || StringUtils.isBlank(adSlot[0]) || StringUtils.isBlank(adSlot[1])) {
-                throw new PreBidException("Invalid adSlot provided");
+        Banner banner = imp.getBanner();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java]: Please make `banner` final.

File: src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java [hunk 2]
```diff
@@ -64,27 +64,41 @@ protected Imp modifyImp(Imp imp, ExtImpPubmatic extImpPubmatic) throws PreBidExc
...
-            final String[] heightStr = adSize[1].trim().split(":");
-            try {
-                width = Integer.valueOf(adSize[0].trim());
-                height = Integer.valueOf(heightStr[0].trim());
-            } catch (NumberFormatException e) {
-                throw new PreBidException("Invalid width or height provided in adSlot");
+                height = isFormatsPresent && banner.getH() == null && banner.getW() == null
+                         ? banner.getFormat().get(0).getH() : banner.getH();
             }
             final Banner updatedBanner = imp.getBanner().toBuilder().w(width).h(height).build();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java]: Minor. The `imp.getBanner()` can be replaced with `banner`.

File: src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java (hunk 1 above)
Comment (pm-deepak-dhage) [reply src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java]: Done.

File: src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java (hunk 2 above)
Comment (pm-deepak-dhage) [reply src/main/java/org/prebid/server/bidder/pubmatic/PubmaticBidder.java]: Done. Used the reference created earlier.

**Pull request merged**